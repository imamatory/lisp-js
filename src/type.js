// @flow

export const getType = data => (data.type || typeof data);
