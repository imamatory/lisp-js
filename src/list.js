// @flow

export const makeEmptyList = () => ({
  type: 'list',
  head: null,
  tail: null,
});

export const isEmptyList = list => (list.head === null && list.tail === null);

export const cons = (head, tail) => ({
  type: 'list',
  head,
  tail,
});

export const car = list => list.head;
export const cdr = list => list.tail;
