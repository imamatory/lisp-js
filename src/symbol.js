// @flow

export const makeSymbol = (value: string) => ({
  type: 'symbol',
  value,
});

export const getValue = sym => sym.value;
