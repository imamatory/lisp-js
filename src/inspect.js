// @flow

import _ from 'lodash';
import { car, cdr, isEmptyList } from './list';
import { getValue } from './symbol';
import { getType } from './type';

const inspect = (data) => {
  const inspectList = (l) => {
    if (isEmptyList(l)) {
      return '';
    }

    const headString = inspect(car(l));
    const tailString = inspectList(cdr(l));
    const listContent = _.trim(`${headString} ${tailString}`);
    return _.isEmpty(tailString) ? listContent : `(${listContent})`;
  };

  const type = getType(data);

  switch (type) {
    case 'list':
      return isEmptyList(data) ? '' : inspectList(data);
    case 'symbol': return getValue(data);
    case 'number': return data;
    case 'string': return `"${data}"`;
    default: return data;
  }
};

export default inspect;
