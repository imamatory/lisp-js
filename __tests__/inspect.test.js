// @flow

import { makeEmptyList, cons } from '../src/list';
import { makeSymbol } from '../src/symbol';
import inspect from '../src/inspect';

test('should render list to string', () => {
  const plus = makeSymbol('+');
  const l1 = cons(plus, cons(3, cons(2, makeEmptyList())));
  const l2 = cons('Test me', makeEmptyList());
  const l3 = cons(plus, cons(l2, l1));

  expect(inspect(l3)).toMatchSnapshot();
});

