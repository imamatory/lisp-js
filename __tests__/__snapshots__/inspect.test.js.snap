// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render list to string 1`] = `"(+ (\\"Test me\\" (+ (3 2))))"`;
